File: submissions/cpp/234.cpp
class Solution {
public:
    bool isPalindrome(ListNode *head) {
        ListNode *first = head, *second = head;
        while (first && first->next) {
            first = first->next->next;
            second = second->next;
        }
        if (first)
            second = second->next;
        second = ReverseList(second);
        while (head && second) {
            if (head->val != second->val)
                return false;
            head = head->next;
            second = second->next;
        }
        return true;
    }

    ListNode *ReverseList(ListNode *head) {
        ListNode *prev = nullptr, *curr = head, *next = head;
        while (curr) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};

File: LeetCode/python3/234.py
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        temp = head
        while temp and temp.next != slow:
            temp = temp.next
        temp.next = None
        curr = slow
        prev = None
        while curr:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        while prev and head:
            if prev.val != head.val:
                return False
            prev = prev.next
            head = head.next
        return True
